File: Scribble/Scribble/TraceViewController.m
//  TraceViewController.m
//
//  Created by Chris Burns on 2/10/14.
//  Copyright 2014 Calgary Scientific Inc. All rights reserved.

#import "TraceViewController.h"

#import <PureWeb/PureWeb.h>
#import <PureWeb/PWTraceLogger.h>

@interface TraceViewController ()

@property (strong, nonatomic) IBOutlet UITextView *traceTextView;

@end

@implementation TraceViewController

- (void)viewDidLoad
{
    //Read in all currently existing messages and add them to the text view
    PWTraceLogger *logger = [PWTraceLogger sharedInstance];
    
    NSArray *messages = [logger getMessages];
    for (NSString *message in messages)
    {
        [self logMessage:message];
    }
    
    //Subscribe to new messages provided by the logger
    [[logger messageLogged] addSubscriber:self action:@selector(logMessage:)];
    
    [super viewDidLoad];
}

- (void)logMessage:(NSString *) message
{
    self.traceTextView.text = [self.traceTextView.text stringByAppendingFormat:@"%@\n", message];
    [self.traceTextView scrollRangeToVisible:NSMakeRange([self.traceTextView.text length], 0)];
}

- (void)dealloc
{
    [[[PWTraceLogger sharedInstance] messageLogged] removeSubscribersForTarget:self];
}

@end

File: Swift-Asteroids/Asteroids/Asteroids/TraceViewController.swift
//
//  TraceViewController.swift
//  Asteroids
//
//  Created by Jonathan Neitz on 2016-02-16.
//  Copyright © 2016 Calgary Scientific Inc. All rights reserved.
//

import UIKit
import PureWeb

class TraceViewController: UIViewController {

    @IBOutlet weak var traceTextView: UITextView!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        let logger = PWTraceLogger.sharedInstance()
        
        let messages = logger?.getMessages()
        
        for message in messages! {
            logMessage(message as! String)
        }
        
        logger?.messageLogged().addSubscriber(self, action: #selector(DDLogger.log(message:)))
    }

    func logMessage(_ msg:String) {
        traceTextView.text = traceTextView.text + msg + "\n";
        traceTextView.scrollRangeToVisible( NSMakeRange(traceTextView.text.characters.count, 0));
    }
    
    deinit {
        let logger = PWTraceLogger.sharedInstance()
        logger?.messageLogged().removeSubscriber(self, action: #selector(DDLogger.log(message:)))
    }
    
}
